### PR TITLE
Implement type and shape inference for OneHotEncoder

### PR DESCRIPTION
**Pull request by jbachurski**
### Description
This PR implements type and shape inference for OneHotEncoder with a basic check. A test is added.

### Motivation and Context
Complete type and shape inference is useful for static model checking, avoiding missing shape propagation in inference.

### CODE REVIEW

File: onnx/test/shape_inference_test.py [hunk 1]
```diff
@@ -8616,6 +8616,31 @@ def test_array_feature_extractor(self) -> None:
                     ],
                 )
 
+    def test_one_hot_encoder(self) -> None:
+        if not ONNX_ML:
```
Comment (justinchuby) [inline onnx/test/shape_inference_test.py:8620]: Prefer a skipIf for this?

Comment (justinchuby) [reply onnx/test/shape_inference_test.py:8620]: So that the test doesn't succeed silently

**Comment by jbachurski** [comment]
PR line count is longer than it actually changes due to indentation changes - `if ONNX_ML:` blocks were replaced with a `skipUnless(ONNX_ML, ...)`.

**Review by xadupre — APPROVED** [review]

**Comment by jbachurski** [comment]
Thanks, ready to merge.

**Pull request merged**